fix(check): Don't generalize variables from an outer scope ...

Marwes opened the pull request:
... when merging with a type signature

This commit tracks all type variables introduced in type signatures and uses this information to emit an error when a variable from a signature is matched to a variable from an outer scope.

Fixes #77

Marwes commented:
This still fails to compile the prelude as there are multiple places where a type signature is unified with a variable from a scope outside the type signature. Would like some feedback for how to fix this, see #77 for some ideas.

brendanzab commented on check/tests/fail.rs:
```diff
@@ -294,3 +294,17 @@ let g x: A a -> () = x
+"#;
+    let result = support::typecheck(text);
+    assert!(result.is_err());
+}
```
Missing new line

brendanzab commented on check/tests/pass.rs:
```diff
@@ -827,3 +826,18 @@ in r1"#;
+"#;
+    let result = support::typecheck(text);
+    assert!(result.is_ok(), "{}", result.unwrap_err());
+}
```
Missing new line

brendanzab commented:
```
Line: 69, Column: 9: Expected the following types to be equal
Expected: m -> m -> m
Found: 88 -> 88 -> 88
3 errors were found during unification:
Could not generalize the variable bound to `m` as the variable was used outside its scope
Could not generalize the variable bound to `m` as the variable was used outside its scope
Could not generalize the variable bound to `m` as the variable was used outside its scope
    let (<>) : m -> m -> m = append
        ^~
```

Well, now that I think about it, the current definition of `make_Monoid` is quite odd:

```
let make_Monoid m =
    let { append, empty } = m

    let (<>) : m -> m -> m = append

    { append, empty, (<>) }
```

What `make_Monoid` should *really* look like is this ('scuse the straw man syntax for implicits #98):

```
let make_Monoid [t] M: [t : Type] -> Monoid t -> { Monoid t | (<>) : t -> t -> t }  =
    let { append, empty } = M

    let (<>) : t -> t -> t = append

    { append, empty, (<>) }
```

Since we have no current way of bringing the `t` binding into scope, we shouldn't be able to explicitly type `(<>)`. I'm not sure if this is the source of the errors we are getting, but there is certainly something unsound in our current typechecker that wasn't getting found.

Marwes commented:
Rebased. Fixed the nits and added some comments to the first commit, fixed a small mistake and commented out the failing type signatures in the other two commits

brendanzab commented on check/src/substitution.rs:
```diff
@@ -294,6 +294,10 @@ impl<T: Substitutable + PartialEq + Clone> Substitution<T> {
     pub fn union(&self, id: &T::Variable, typ: &T) -> Result<(), ()>
         where T::Variable: Clone
     {
+        // Nothing needs to be done if both are the same variable already (also prevents the occurs check from failing)
```
> also prevents the occurs check from failing

What does this mean? Would `occurs` panic otherwise?

Marwes replied:
No it would fail Typecheck ingredients with `123 occurs in 123` which is obviously nonsense. Unifying with itself should always succeed

brendanzab replied:
Aha! Thanks!

brendanzab commented on check/src/unify_type.rs:
```diff
@@ -375,6 +384,109 @@ fn walk_move_types2<'a, I, F, T>(mut types: I, replaced: bool, output: &mut Vec<
+impl<'a, 'e> Unifier<State<'a>, TcType> for Merge<'e> {
+    fn report_error(unifier: &mut UnifierState<Self>,
+                    error: UnifyError<TcType, TypeError<Symbol>>) {
+        unifier.unifier.errors.error(error);
```
The stutteriness is strong with this one... :3

Marwes replied:
And naming is hard :(
